### PR TITLE
Public ids threadsafe fix

### PR DESCRIPTION
**Pull request by andrewconner**
@2is10 Guava / Scala has this weird bug when storing `java.util.Long`s (Scala tries to optimize it and unbox them, but that blows up the cache at runtime). Without the cache seems fast enough, but I'll keep an eye on it.

### CODE REVIEW

File: kifi-backend/modules/common/app/com/keepit/common/crypto/ModelWithPublicId.scala [hunk 1]
```diff
@@ -58,9 +59,21 @@ trait ModelWithPublicIdCompanion[T <: ModelWithPublicId[T]] {
   */
   protected[this] val publicIdIvSpec: IvParameterSpec
 
+  private def threadSafeDecrypt(id: Long, config: PublicIdConfiguration): Try[Long] = {
+    synchronized {
```
Comment (ymatsuda) [inline kifi-backend/modules/common/app/com/keepit/common/crypto/ModelWithPublicId.scala]: Which part is thread-unsafe? If it is Cipher, why don't you synchronized just on cipher.doFinal in Aes64BitCipher?

File: kifi-backend/modules/common/app/com/keepit/common/crypto/ModelWithPublicId.scala [hunk 2]
```diff
@@ -58,9 +59,21 @@ trait ModelWithPublicIdCompanion[T <: ModelWithPublicId[T]] {
...
+      Try(config.aes64bit(publicIdIvSpec).decrypt(id))
+    }
```
Comment (2is10) [inline kifi-backend/modules/common/app/com/keepit/common/crypto/ModelWithPublicId.scala]: This looks correct, but synchronizing on the `ModelWithPublicIdCompanion` relies on knowledge of how `PublicIdConfiguration` manages `Aes64BitCipher` instance lifetimes.

We could localize that knowledge and the solution to the thread-safety concern better by making `PublicIdConfiguration` handle both creation/reuse _and_ synchronization. The `PublicIdConfiguration` API might then look like:

``` scala
config.encode(publicIdIvSpec, id)  // or .encrypt
config.decode(publicIdIvSpec, id)  // or .decrypt
```

File: kifi-backend/modules/common/app/com/keepit/common/crypto/ModelWithPublicId.scala (hunk 1 above)
Comment (andrewconner) [inline kifi-backend/modules/common/app/com/keepit/common/crypto/ModelWithPublicId.scala]: True, thanks

File: kifi-backend/modules/common/app/com/keepit/common/crypto/ModelWithPublicId.scala (hunk 2 above)
Comment (andrewconner) [inline kifi-backend/modules/common/app/com/keepit/common/crypto/ModelWithPublicId.scala]: That's better, will do

**Comment by andrewconner** [comment]
@ymatsuda Simplified based on your feedback. If this is a bottleneck later, I can complicate it to make it faster

**Comment by 2is10** [comment]
:+1:

**Pull request merged**